library/VTKExtensions/Rendering/vtkF3DRenderer.cxx

```diff
@@ -498,131 +502,152 @@ void vtkF3DRenderer::SetHDRIFile(const std::string& hdriFile)
-      }
+    // Check spherical harmonics cache
+    std::string shCachePath = this->CachePath + "/" + hash + "/sh.vtt";
+    bool shCacheExists = vtksys::SystemTools::FileExists(shCachePath, true);
+    if (shCacheExists)
+    {
+      vtkNew<vtkXMLTableReader> reader;
+      reader->SetFileName(shCachePath.c_str());
+      reader->Update();
 
-      // Check specular cache
-      std::string specCachePath = this->CachePath + "/" + hash + "/specular.vtm";
-      bool specCacheExists = vtksys::SystemTools::FileExists(specCachePath, true);
-      if (specCacheExists)
+      this->SphericalHarmonics = vtkFloatArray::SafeDownCast(reader->GetOutput()->GetColumn(0));
+    }
+
+    // Check specular cache
+    std::string specCachePath = this->CachePath + "/" + hash + "/specular.vtm";
+    bool specCacheExists = vtksys::SystemTools::FileExists(specCachePath, true);
+    if (specCacheExists)
+    {
+      // TODO: this->EnvMapPrefiltered should be a smart pointer in VTK
+      if (this->EnvMapPrefiltered)
       {
-        vtkF3DCachedSpecularTexture* spec = vtkF3DCachedSpecularTexture::New();
-        spec->SetFileName(specCachePath.c_str());
-        this->EnvMapPrefiltered = spec;
+        this->EnvMapPrefiltered->Delete();
       }
 
-      this->GetEnvMapPrefiltered()->HalfPrecisionOff();
-#endif
-
-      // HDRI OpenGL
-      this->UseImageBasedLightingOn();
-      this->SetEnvironmentTexture(hdriTexture);
-
-      // Skybox OpenGL
-      this->Skybox->SetProjection(vtkSkybox::Sphere);
-      this->Skybox->SetTexture(hdriTexture);
+      vtkF3DCachedSpecularTexture* spec = vtkF3DCachedSpecularTexture::New();
+      spec->SetFileName(specCachePath.c_str());
+      this->EnvMapPrefiltered = spec;
+    }
 
-      // First version of VTK including the version check (and the feature used)
-#if VTK_VERSION_NUMBER >= VTK_VERSION_CHECK(9, 0, 20200527)
-      this->Skybox->GammaCorrectOn();
+    this->GetEnvMapPrefiltered()->HalfPrecisionOff();
 #endif
 
-      // skybox orientation
-      double front[3];
-      vtkMath::Cross(this->RightVector, this->UpVector, front);
-      this->Skybox->SetFloorPlane(this->UpVector[0], this->UpVector[1], this->UpVector[2], 0.0);
-      this->Skybox->SetFloorRight(front[0], front[1], front[2]);
-
-      this->AddActor(this->Skybox);
-
-      // build HDRI textures and spherical harmonics
-      this->Render();
+    // HDRI OpenGL
+    this->UseImageBasedLightingOn();
+    this->SetEnvironmentTexture(hdriTexture);
 
 #if VTK_VERSION_NUMBER >= VTK_VERSION_CHECK(9, 2, 20221220)
-      // Create LUT cache file
-      if (!lutCacheExists)
+    // Create LUT cache file
+    if (!lutCacheExists)
+    {
+      vtkPBRLUTTexture* lut = this->GetEnvMapLookupTable();
+      if (!lut->GetTextureObject())
+      {
+        this->Render();
+        lut = this->GetEnvMapLookupTable();
+      }
+      if (!lut->GetTextureObject())
+      {
+        F3DLog::Print(F3DLog::Severity::Error, "Cannot recover a HDRI texture environment map");
```

Comment:
```suggestion
        F3DLog::Print(F3DLog::Severity::Error, "Cannot compute the PBR look-up texture");
```

Reply:
actually, should this be an assert ? can this be empty after a render ?

Reply:
It's not supposed to be empty, let's put an assert :)